opte/src/engine/icmpv6.rs

```diff
@@ -346,18 +349,141 @@ impl HairpinAction for RouterAdvertisement {
     }
 }
 
+// Check if an ICMPv6 message is a valid Neighbor Solicitation
+//
+// See https://www.rfc-editor.org/rfc/rfc4861.html#section-7.1.1 for details on
+// the validations performed.
+//
+// Return the target address and whether the Link-Layer address option is
+// included, the data from the Neighbor Solicitation.
+fn validate_neighbor_solicitation(
+    rdr: &mut PacketReader<Parsed, ()>,
+    metadata: &Ipv6Meta,
+) -> Result<(Ipv6Addr, bool), GenErr> {
+    // First, check if this is in fact a NS message.
+    let smol_src = IpAddress::Ipv6(metadata.src.into());
+    let smol_dst = IpAddress::Ipv6(metadata.dst.into());
+    let body = rdr.copy_remaining();
+    let src_pkt = Icmpv6Packet::new_checked(&body)?;
+    let icmp =
+        Icmpv6Repr::parse(&smol_src, &smol_dst, &src_pkt, &Csum::ignored())?;
+
+    let (target_addr, has_ll_option) = match icmp {
+        Icmpv6Repr::Ndisc(NdiscRepr::NeighborSolicit {
+            lladdr,
+            target_addr,
+        }) => (Ipv6Addr::from(target_addr), lladdr.is_some()),
+        _ => {
+            // We should never hit this case because the predicate
+            // should have verified that we are dealing with a
+            // Neighbor Solicitation. However, programming error could
+            // cause this to happen -- let's not take any chances.
+            return Err(GenErr::Unexpected(format!(
+                "expected a NDP Neighbor Solicitation, got {} {}",
+                src_pkt.msg_type(),
+                src_pkt.msg_code()
+            )));
+        }
+    };
+
+    // The target cannot be a multicast address.
+    if target_addr.is_multicast() {
+        return Err(GenErr::Unexpected(String::from(
+            "Received NS with multicast target address, dropping.",
```

Comment:
We don't have to include the "dropping" in error messages. They bubble up as a `ProcessResult::Drop`.